aloe/parser.py

```diff
@@ -7,121 +7,112 @@
+        else:
+            raise ValueError("Must provide either string or filename.")
+
+        self.line_number = 0
+
+    def read(self):
+        """Read the next token from the source."""
+        self.line_number += 1
+        location = {'line': self.line_number}
+        line = self.io.readline()
+        return Token(
+            (GherkinLine(line, self.line_number) if line else line),
+            location
+        )
 
-    def __init__(self, parent, s, loc, filename=None):
-        self.parent = parent
-        self._file = filename
-        self.line = lineno(loc, s)
-        self.col = col(loc, s)
 
-    def __str__(self):
-        return u'{file}:{line}'.format(file=self.file,
-                                       line=self.line)
+class LanguageTokenMatcher(TokenMatcher):
+    """A token matcher that actually remembers the language it's set to."""
 
-    def __repr__(self):
-        return str(self)
+    def __init__(self, dialect_name='en'):
+        self.actual_dialect_name = dialect_name
+        TokenMatcher.__init__(self, dialect_name=dialect_name)
 
-    @property
-    def file(self):
-        """
-        Return the relative path to the file.
-        """
+    def _change_dialect(self, dialect_name, location=None):
+        """Force the dialect name given in the constructor."""
+        TokenMatcher._change_dialect(
+            self, self.actual_dialect_name, location=location)
 
-        if self._file:
-            return os.path.relpath(self._file)
-        elif self.parent:
-            if self.parent.feature.described_at is self:
-                return None
-            return self.parent.feature.described_at.file
-        else:
-            return None
 
-    @file.setter
-    def file(self, value):
-        """Set the file."""
-        self._file = value
+def cell_values(row):
+    """Extract cell values from a table header or row."""
+
+    return tuple(cell['value'] for cell in row['cells'])
 
 
 class Node(object):
     """
-    A parse node
+    A base parse node.
     """
 
-    def __init__(self, s, loc, tokens):
-        self.described_at = ParseLocation(self, s, loc)
-        self.text = pyparsing.line(loc, s)
+    def __init__(self, parsed, filename=None):
+        """Construct the node from Gherkin parse results."""
+        self.line = parsed['location']['line']
+        self.col = parsed['location']['column']
+        self.filename = filename
+
+    @property
+    def feature(self):
+        """The feature this node ultimately belongs to."""
+
+        raise NotImplementedError
+
+    @property
+    def text(self):
+        """The text for this node."""
+
+        raise NotImplementedError
 
     def represented(self, indent=0, annotate=True):
-        """
-        Return a representation of the node.
-        """
+        """A representation of the node."""
 
-        s = u' ' * indent + self.text.strip()
+        result = ' ' * indent + self.text.strip()
 
         if annotate:
```

Comment:
We can probably remove annotate. It's only used in one place, and it could be replaced by something better, which would also fix the issue Kane has a PR open for. It dates from Lettuce.